sql: reduce miscellaneous allocations

yuzefovich opened the pull request:
See each commit for details.

Combined improvement:
```
name                   old time/op    new time/op    delta
LastWriteWinsInsert-8    1.42ms ±16%    1.44ms ±13%    ~     (p=0.820 n=20+20)

name                   old alloc/op   new alloc/op   delta
LastWriteWinsInsert-8     267kB ± 2%     265kB ± 2%    ~     (p=0.057 n=19+20)

name                   old allocs/op  new allocs/op  delta
LastWriteWinsInsert-8     1.95k ± 0%     1.91k ± 0%  -2.38%  (p=0.000 n=17+18)
```

Epic: CRDB-39063.

yuzefovich committed:
sql: avoid an allocation in internal executor

```
name                   old time/op    new time/op    delta
LastWriteWinsInsert-8    1.44ms ±14%    1.46ms ±14%    ~     (p=0.640 n=20+20)

name                   old alloc/op   new alloc/op   delta
LastWriteWinsInsert-8     267kB ± 1%     268kB ± 1%    ~     (p=0.667 n=19+20)

name                   old allocs/op  new allocs/op  delta
LastWriteWinsInsert-8     1.96k ± 0%     1.95k ± 0%  -0.18%  (p=0.007 n=17+17)
```

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/125724)
<!-- Reviewable:end -->

yuzefovich commented:
```
name                   old time/op    new time/op    delta
LastWriteWinsInsert-8    1.44ms ±14%    1.46ms ±14%    ~     (p=0.640 n=20+20)

name                   old alloc/op   new alloc/op   delta
LastWriteWinsInsert-8     267kB ± 1%     268kB ± 1%    ~     (p=0.667 n=19+20)

name                   old allocs/op  new allocs/op  delta
LastWriteWinsInsert-8     1.96k ± 0%     1.95k ± 0%  -0.18%  (p=0.007 n=17+17)
```

yuzefovich committed:
physicalplan: avoid a redundant allocation in ivarRemapper

We often have an identity "indexVarMap", and in those cases we can avoid
allocating fresh IndexedVars in the ivarRemapper.

```
name                   old time/op    new time/op    delta
LastWriteWinsInsert-8    1.45ms ±17%    1.43ms ±17%    ~     (p=0.529 n=20+20)

name                   old alloc/op   new alloc/op   delta
LastWriteWinsInsert-8     267kB ± 2%     266kB ± 1%    ~     (p=0.079 n=20+19)

name                   old allocs/op  new allocs/op  delta
LastWriteWinsInsert-8     1.95k ± 0%     1.93k ± 0%  -1.31%  (p=0.000 n=15+17)
```

Release note: None

yuzefovich committed:
colmem: move one field from Allocator to AccountingHelper

The goal is to make Allocator store less state so that more objects can
be reused between multiple operators in the following commit.

Release note: None

yuzefovich committed:
colbuilder: share streaming allocators

This commit makes it so that simple streaming operators that are using
the shared "streaming mem account" also use the same "streaming"
Allocator object. Given that each allocator stores no state other than
the memory account, and previously all of these usages were already
using the same memory account, this should be a safe change. The
streaming allocator object is allocated lazily (the streaming memory
account is also now allocated lazily).

```
name                   old time/op    new time/op    delta
LastWriteWinsInsert-8    1.45ms ±15%    1.42ms ±18%    ~     (p=0.495 n=20+20)

name                   old alloc/op   new alloc/op   delta
LastWriteWinsInsert-8     267kB ± 1%     265kB ± 2%  -0.79%  (p=0.008 n=19+20)

name                   old allocs/op  new allocs/op  delta
LastWriteWinsInsert-8     1.93k ± 0%     1.91k ± 0%  -1.18%  (p=0.000 n=17+17)
```

Release note: None

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice!

> Reviewed 1 of 1 files at r1, 1 of 1 files at r2, 3 of 3 files at r3, 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125724#-:-O-bXbob2IuuXu6Ebljj:b-8gk8uj)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
TFTR!

bors r+
